File: backend/app/api/stages.py
from __future__ import annotations

from fastapi import APIRouter

from app.services.stages_service import (
    complete_stage_bc,
    complete_stage_ppo,
    get_stage_detail,
    list_stages,
    reset_stage_data,
    resume_stage_ppo,
    resume_stage_bc,
    start_stage_ppo,
    start_stage_bc,
    stop_stage_bc,
    stop_stage_ppo,
)

router = APIRouter()


@router.get("/api/stages")
def get_stages() -> list[dict]:
    return list_stages()


@router.get("/api/cache/stages/reset")
def reset_stage_cache() -> dict:
    return {"ok": True}


@router.get("/api/stages/{stage_id}")
def get_stage(stage_id: int) -> dict:
    return get_stage_detail(stage_id)


@router.get("/api/stages/{stage_id}/reset")
def reset_stage(stage_id: int) -> dict:
    try:
        return reset_stage_data(stage_id)
    except ValueError as e:
        return {"ok": False, "error": str(e), "stage_id": stage_id}


@router.get("/api/stages/{stage_id}/bc/start")
def start_stage_bc_api(stage_id: int) -> dict:
    try:
        return start_stage_bc(stage_id)
    except ValueError as e:
        return {"ok": False, "error": str(e), "stage_id": stage_id}
    except Exception as e:
        return {"ok": False, "error": f"start bc failed: {e}", "stage_id": stage_id}


@router.get("/api/stages/{stage_id}/bc/resume")
def resume_stage_bc_api(stage_id: int) -> dict:
    try:
        return resume_stage_bc(stage_id)
    except ValueError as e:
        return {"ok": False, "error": str(e), "stage_id": stage_id}
    except Exception as e:
        return {"ok": False, "error": f"resume bc failed: {e}", "stage_id": stage_id}


@router.get("/api/stages/{stage_id}/ppo/start")
def start_stage_ppo_api(stage_id: int) -> dict:
    try:
        return start_stage_ppo(stage_id)
    except ValueError as e:
        return {"ok": False, "error": str(e), "stage_id": stage_id}
    except Exception as e:
        return {"ok": False, "error": f"start ppo failed: {e}", "stage_id": stage_id}


@router.get("/api/stages/{stage_id}/ppo/resume")
def resume_stage_ppo_api(stage_id: int) -> dict:
    try:
        return resume_stage_ppo(stage_id)
    except ValueError as e:
        return {"ok": False, "error": str(e), "stage_id": stage_id}
    except Exception as e:
        return {"ok": False, "error": f"resume ppo failed: {e}", "stage_id": stage_id}


@router.get("/api/stages/{stage_id}/bc/stop")
def stop_stage_bc_api(stage_id: int) -> dict:
    try:
        return stop_stage_bc(stage_id)
    except Exception as e:
        return {"ok": False, "error": f"stop bc failed: {e}", "stage_id": stage_id}


@router.get("/api/stages/{stage_id}/ppo/stop")
def stop_stage_ppo_api(stage_id: int) -> dict:
    try:
        return stop_stage_ppo(stage_id)
    except Exception as e:
        return {"ok": False, "error": f"stop ppo failed: {e}", "stage_id": stage_id}


@router.get("/api/stages/{stage_id}/bc/complete")
def complete_stage_bc_api(stage_id: int) -> dict:
    try:
        return complete_stage_bc(stage_id)
    except ValueError as e:
        return {"ok": False, "error": str(e), "stage_id": stage_id}
    except Exception as e:
        return {"ok": False, "error": f"complete bc failed: {e}", "stage_id": stage_id}


@router.get("/api/stages/{stage_id}/ppo/complete")
def complete_stage_ppo_api(stage_id: int) -> dict:
    try:
        return complete_stage_ppo(stage_id)
    except ValueError as e:
        return {"ok": False, "error": str(e), "stage_id": stage_id}
    except Exception as e:
        return {"ok": False, "error": f"complete ppo failed: {e}", "stage_id": stage_id}

File: frontend/src/api/stages.js
const DEFAULT_API_BASE = "http://127.0.0.1:8000";

export async function fetchStages(signal) {
  const apiBase = import.meta.env.VITE_API_BASE || DEFAULT_API_BASE;
  const res = await fetch(`${apiBase}/api/stages`, { signal });
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  const data = await res.json();
  return Array.isArray(data) ? data : [];
}

export async function fetchStageDetail(stageId, signal) {
  const apiBase = import.meta.env.VITE_API_BASE || DEFAULT_API_BASE;
  const res = await fetch(`${apiBase}/api/stages/${encodeURIComponent(stageId)}`, { signal });
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  const data = await res.json();
  return data && typeof data === "object" ? data : null;
}

export async function resetStage(stageId) {
  const apiBase = import.meta.env.VITE_API_BASE || DEFAULT_API_BASE;
  const res = await fetch(`${apiBase}/api/stages/${encodeURIComponent(stageId)}/reset`);
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  const data = await res.json();
  if (!data || typeof data !== "object") throw new Error("invalid response");
  if (!data.ok) throw new Error(data.error || "reset failed");
  return data;
}

export async function startStageBc(stageId) {
  const apiBase = import.meta.env.VITE_API_BASE || DEFAULT_API_BASE;
  const res = await fetch(`${apiBase}/api/stages/${encodeURIComponent(stageId)}/bc/start`);
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  const data = await res.json();
  if (!data || typeof data !== "object") throw new Error("invalid response");
  if (!data.ok) throw new Error(data.error || "start bc failed");
  return data;
}

export async function stopStageBc(stageId) {
  const apiBase = import.meta.env.VITE_API_BASE || DEFAULT_API_BASE;
  const res = await fetch(`${apiBase}/api/stages/${encodeURIComponent(stageId)}/bc/stop`);
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  const data = await res.json();
  if (!data || typeof data !== "object") throw new Error("invalid response");
  if (!data.ok) throw new Error(data.error || "stop bc failed");
  return data;
}

export async function resumeStageBc(stageId) {
  const apiBase = import.meta.env.VITE_API_BASE || DEFAULT_API_BASE;
  const res = await fetch(`${apiBase}/api/stages/${encodeURIComponent(stageId)}/bc/resume`);
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  const data = await res.json();
  if (!data || typeof data !== "object") throw new Error("invalid response");
  if (!data.ok) throw new Error(data.error || "resume bc failed");
  return data;
}

export async function completeStageBc(stageId) {
  const apiBase = import.meta.env.VITE_API_BASE || DEFAULT_API_BASE;
  const res = await fetch(`${apiBase}/api/stages/${encodeURIComponent(stageId)}/bc/complete`);
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  const data = await res.json();
  if (!data || typeof data !== "object") throw new Error("invalid response");
  if (!data.ok) throw new Error(data.error || "complete bc failed");
  return data;
}

export async function startStagePpo(stageId) {
  const apiBase = import.meta.env.VITE_API_BASE || DEFAULT_API_BASE;
  const res = await fetch(`${apiBase}/api/stages/${encodeURIComponent(stageId)}/ppo/start`);
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  const data = await res.json();
  if (!data || typeof data !== "object") throw new Error("invalid response");
  if (!data.ok) throw new Error(data.error || "start ppo failed");
  return data;
}

export async function stopStagePpo(stageId) {
  const apiBase = import.meta.env.VITE_API_BASE || DEFAULT_API_BASE;
  const res = await fetch(`${apiBase}/api/stages/${encodeURIComponent(stageId)}/ppo/stop`);
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  const data = await res.json();
  if (!data || typeof data !== "object") throw new Error("invalid response");
  if (!data.ok) throw new Error(data.error || "stop ppo failed");
  return data;
}

export async function resumeStagePpo(stageId) {
  const apiBase = import.meta.env.VITE_API_BASE || DEFAULT_API_BASE;
  const res = await fetch(`${apiBase}/api/stages/${encodeURIComponent(stageId)}/ppo/resume`);
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  const data = await res.json();
  if (!data || typeof data !== "object") throw new Error("invalid response");
  if (!data.ok) throw new Error(data.error || "resume ppo failed");
  return data;
}

export async function completeStagePpo(stageId) {
  const apiBase = import.meta.env.VITE_API_BASE || DEFAULT_API_BASE;
  const res = await fetch(`${apiBase}/api/stages/${encodeURIComponent(stageId)}/ppo/complete`);
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  const data = await res.json();
  if (!data || typeof data !== "object") throw new Error("invalid response");
  if (!data.ok) throw new Error(data.error || "complete ppo failed");
  return data;
}
